### DISCUSSION TITLE
Let's contribute to already existing solutions instead of reinventing the wheel

### DISCUSSION THREAD

**Discussion by kromka-chleba** [microsoft/winget-cli#323] [General]
# Description of the new feature/enhancement

There already exist full-featured package managers on GNU/Linux and other UNIX-like operating systems. Why waste time on reinventing already existing features, encounter common mistakes, when you could port one to Windows?

This way you avoid making yet another standard, potentially abandoned and forgotten later, lacking basic features like uninstalling at the current stage of the project.

There are two especially good package managers:

Guix
http://guix.gnu.org/

Nix
https://nixos.org/

By using one of these you gain building and deploying software in reproducible and declarative manner.
You can read more here:
http://guix.gnu.org/manual/en/html_node/Features.html
https://nixos.org/nix/

# Why even bother?

- by using already existing standard, you gain more **support and popularity**

- already existing solutions are **more mature**

- not wasting time on doing what's already done and concentrating on making it better

- better integration with WSL, using **one solutions to manage both systems**

- providing a **convenient environment for developers** - defining reproducible software packages, preventing DLL hell, building from source code

# Proposed technical implementation details (optional)

For porting Guix to Windows you would have to:

- port Guile (Scheme dialect) to Windows
https://www.gnu.org/software/guile/
It is a programming language Guix is implemented in, also packages are written in it.
Some work with mingw-64 is already done, it shouldn't be that hard to port.
It uses a WM so it is easier to port:
https://www.gnu.org/software/guile/manual/guile.html#A-Virtual-Machine-for-Guile
It is possible to combine Guile with C and add support for other programming languages, for example you could add PowerShell and allow making packages and extending the package manager in it:
https://www.gnu.org/software/guile/manual/guile.html#Supporting-Multiple-Languages

- add support for Windows for already existing packages written in Guile

- or just use WSL instead of porting

As for Nix, I don't know, because I'm not that familiar with it as with Guix.

**Comment by MatejKafka** [comment]
From what I understand, `winget` allows developers to reuse their `.msi` and `.msix` installers, unlike both your proposed package managers, which have their own (uncommon) format. Also, don't know about Guix, but Nix way of building apps is quite different from the way Windows apps are currently built, with custom patched compilers and declarative funcional language, which would require large amount of effort for both Microsoft and app developers to adapt.

Yet another issue - Nix sandboxes its packages in a way that would probably require major codebase changes for most apps, and afaik is not really friendly to closed-source packages, which are quite common on Windows.

In summary, while I'd definitely support Nix for Windows, as the concept is really cool, there are major issues with practical implementation that make it unlikely.

**Comment by kromka-chleba** [comment]
@MatejKafka 
For Guix or Nix it doesn't matter if a package is proprietary or not, because of the functional way of dealing with software - Nix doesn't care if it's source code, text file, image or an executable (hypothetically .exe, .msi, etc.) - all you got is input data, procedures to apply,  and output data.
Nix even has some proprietary packages, like Steam
https://nixos.org/nixos/packages.html?attr=steam&channel=nixos-20.03&query=steam

And there's no a format on Guix/Nix - there's the store - a directory where "derivations" (effects of building a package) are kept - it doesn't matter if the package was compiled by user or developer, packages can be just downloaded to the store.

http://guix.gnu.org/manual/en/html_node/The-Store.html#The-Store
http://guix.gnu.org/manual/en/html_node/Derivations.html#Derivations

Using legacy formats such as .msi is a good option for providing backward compatibility but if it is going to be the only format used to install software, then winget isn't really a package manager, like the other user noted #223 , but just a downloader.

As for the major codebase change, for years Microsoft have failed to provide unified way of delivering software to Windows users - some developers use .exe installers, other use .msi, zip archives, there's also appx and the Windows store barely anyone uses, there's the chocolatey package manager. If winget isn't going to innovate, but will end up as a downloader for .msi files, then no one will use it. On the other hand if winget was a fully featured package manager and provided real advantages over previous ways of installing software, developers would switch without forcing them to.

A functional package manager such as Nix could reuse any of these formats by simply executing them and keeping in the store, but it isn't limited to just it - custom package definitions for building software from source code could be written, if developers wished to do so.

But looking at how many people dislike the idea, I think it's better to grab some popcorn and watch how developers reinvent the wheel, port it to GNU/Linux for Azure and then abandon just like Windows Phone.

**Comment by ghost** [comment]
@kromka-chleba The challenges involved with creating a package manager for Windows are primarily legal ones, not technical ones. The current state of copyright law simply does not allow for public package repositories like the ones you're used to on Linux to exist on Windows. You can check out #340 for more details.

**Comment by natesubra** [comment]
[standards](https://imgs.xkcd.com/comics/standards.png)

That said, there are plenty of options for managing packages on windows:
- My preference -> [scoop](https://github.com/lukesampson/scoop) 
- [Chocolatey](https://chocolatey.org)
- [Ketarin](https://ketarin.org)
- [just-install](https://just-install.github.io/)

It's still great to see "competition", or some type of official support in this area. Winget so far seems like almost a 1 for 1 replica of scoop minus some other features (ability to run a repo straight out of github, portable application installation, etc). I use scoop, and I'll probably use winget. I don't see competition in this space as problematic. I also don't think it's helpful to compare the Windows software ecosystem to 'nix based ecosystems. While they are alike and we can compare some attributes, it's definitely apples to oranges, from licensing, to packaging.

**Reply by g0t4** [reply]
Agreed, competition is great. 

- `winget` doesn't need to do and be everything that `chocolatey`, `scoop`, etc do... I
- I see them as complimentary tools
- winget is bundled (in Win11 last I checked) 
    - which would make it a simple way to bootstrap `chocolatey`, `scoop`, etc
    - and, installing from `winget` bolsters trust given MSFT's stamp of approval, thus helping people overcome concerns about safety which are natural (and very valid) in this space.
- `winget` will be a **gateway drug** to `chocolatey`, `scoop`, etc 
    - for novices to rapidly experience the value of package managers (automation, reproducibility)
    - to graduate to tools like `chocolatey` when they run into apps not installable via `winget` that you can readily get from `chocolatey`, case in point `sysinternals`
- I haven't found it yet, if it exists, but it would be nice to add `chocolatey` via `winget source add`!
    - Similar to how `ChocolateyGet` plugs `chocolatey` into `OneGet`/`PackageManagement` as a `PackageProvider`!

**Comment by Amondale** [comment]
I'm all for anything that doesn't bundle malware/adware/spyware into Open Source. I also happen to trust MSFT more than most Open Source consortiums for doing so. No way I'm (or most smaller dev shops are) going to read every line of 16 000 lines of code to see if a cool bundle is referencing a routine to catalog or crypto-lock my hard drive,  

Wait, I'm all for anything that lets me quickly rebuild a dev instance if it does get crypto-locked, which I believe is one of the primary features of WinGet for those that don't exclusively use LAMP dev environments. Docker and even WSL2 environments, anything "container-based" need to be spun up quickly and reliably as the first requirement.

Let's see how WinGet evolves, as WSL2 acquires a GUI; I suspect MSFT pays a good bit of attention to these threads on GitHub.
